purge_history_api.rst: add example script

rubo77 opened the pull request.

rubo77 commented:
maybe we should also add an example, how to select an event_id at a certain date with an SQL command, so we can prune a room regularly

rubo77 commented on docs/admin_api/purge_history_api.rst:
```diff
@@ -13,3 +13,50 @@ The API is simply:
+    # $ synctl start
+
+    # This could be set, so you don't need to prune every time after deleting some rows:
+    # $ sqlite3 homeserver.db "PRAGMA auto_vacuum = FULL;"
```
It seems like `auto_vacuum = FULL` does not have any effect?

rubo77 commented:
this all works fine, apart from sometimes getting "the database is locked" so the script doesn't get some data. How can I prevent that?

rubo77 commented:
I fixed the database is locked problem with a pragma busy timeout, so this can be merged now

kyrias commented:
Would be nice if this supported Postgres too, but you should probably recommend stopping synapse before running the query against the sqlite3 DB, because multiple writers to an sqlite3 DB is a recipe for corruption.

rubo77 commented:
I thought the busy_timeout will sort this out further?
And if you call via API,  the calls are rolled out one after another

rryan commented:
This was helpful, thanks! BTW the pragma busy timeout adds an additional line of output to BUFFER which prevents awk from properly selecting the event ID.

rubo77 commented:
@rryan do you mean we have to change something more? or do you just explain, why I used this?

    EVENT_ID=$(echo $BUFFER|awk '{print $2}')

4nd3r commented:
FYI #1621

richvdh commented:
The API has changed now, so I'm afraid this is out of date.

rubo77 commented:
@richvdh: I think the changes are not relevant to my script

though it would be wise to add 

    POSTDATA='{"delete_local_events":"true"}'
    curl -d $POSTDATA

to my script to really delete local events if desired

rubo77 commented:
I added hints how to solve this in Postgres too.

And I added the command to really delete local events and messages, which is what users want to achieve mostly i think.

rubo77 commented:
changelog added and commits squashed.

rubo77 commented:
You could easily adapt the script to delete all rooms with a loop like

```
ROOMS=$(sudo -u postgres psql -t -A --dbname="$DBNAME" --command="SELECT room_id FROM rooms;" 2>/dev/null)

for ROOM_NAME in $ROOMS; do
  ...

```

rubo77 commented:
But I suggest to only purge old history in rooms with many users, for example more than 10 users:

`select * from (select count(*) as numberofusers, room_id from current_state_events where type ='m.room.member' group by room_id) as q left join room_aliases a on q.room_id=a.room_id where q.numberofusers >= 10 order by numberofusers desc`

richvdh commented:
sorry, I don't think we're going to be able to accept a script of this complexity into the official documentation; we won't have the resources to maintain it.

We could consider putting something into `contrib` if you like.

rubo77 commented:
@richvdh ok, should I refactor the PR, so it adds the script to contrib/ ?

rubo77 commented:
I created a new PR in https://github.com/matrix-org/synapse/pull/4155 with a much enhanced script, that loops over more rooms and waits for one purge to finish until the next one starts
